### ISSUE TITLE
attributes at end of scope

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#5824]
Migrated from [rt.perl.org#16080](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=16080) (status was 'resolved')

Searchable as RT16080$

**Comment by p5pRT** [comment]
### From @lizmat

#### Created by @lizmat



$ perl \-e '\{my $foo : shared\}'
Invalid separator character '\}' in attribute list at \-e line 1\, near "$foo 
: shared"
syntax error at \-e line 1\, near "$foo : shared"
Execution of \-e aborted due to compilation errors\.


Found this problem when trying to do a&#8203;:

&nbsp;   BEGIN \{our $foo : shared\}

of course&#8203;:

&nbsp;   BEGIN \{our $foo : shared;\}

works ok\.\.\.

<details><summary>Perl Info</summary>

```

Flags:
     category=core
     severity=low

Site configuration information for perl v5.8.0:

Configured by liz at Thu Aug  1 12:25:00 CEST 2002.

Summary of my perl5 (revision 5.0 version 8 subversion 0) configuration:
   Platform:
     osname=linux, osvers=2.4.18, archname=i686-linux-thread-multi
     uname='linux echt.dijkmat.nl 2.4.18 #8 smp mon mar 25 22:28:36 cet 
2002 i686 unknown '
     config_args='-Doptimize=-g -de -Dusethreads'
     hint=recommended, useposix=true, d_sigaction=define
     usethreads=define use5005threads=undef useithreads=define 
usemultiplicity=define
     useperlio=define d_sfio=undef uselargefiles=define usesocks=undef
     use64bitint=undef use64bitall=undef uselongdouble=undef
     usemymalloc=n, bincompat5005=undef
   Compiler:
     cc='cc', ccflags ='-D_REENTRANT -D_GNU_SOURCE -DDEBUGGING 
-fno-strict-aliasing -I/usr/local/include -D_LARGEFILE_SOURCE 
-D_FILE_OFFSET_BITS=64 -I/usr/include/gdbm',
     optimize='-g',
     cppflags='-D_REENTRANT -D_GNU_SOURCE -DDEBUGGING -fno-strict-aliasing 
-I/usr/local/include -I/usr/include/gdbm'
     ccversion='', gccversion='2.96 20000731 (Red Hat Linux 7.1 2.96-98)', 
gccosandvers=''
     intsize=4, longsize=4, ptrsize=4, doublesize=8, byteorder=1234
     d_longlong=define, longlongsize=8, d_longdbl=define, longdblsize=12
     ivtype='long', ivsize=4, nvtype='double', nvsize=8, Off_t='off_t', 
lseeksize=8
     alignbytes=4, prototype=define
   Linker and Libraries:
     ld='cc', ldflags =' -L/usr/local/lib'
     libpth=/usr/local/lib /lib /usr/lib
     libs=-lnsl -lgdbm -ldb -ldl -lm -lpthread -lc -lcrypt -lutil
     perllibs=-lnsl -ldl -lm -lpthread -lc -lcrypt -lutil
     libc=/lib/libc-2.2.4.so, so=so, useshrplib=false, libperl=libperl.a
     gnulibc_version='2.2.4'
   Dynamic Linking:
     dlsrc=dl_dlopen.xs, dlext=so, d_dlsymun=undef, ccdlflags='-rdynamic'
     cccdlflags='-fpic', lddlflags='-shared -L/usr/local/lib'

Locally applied patches:



@INC for perl v5.8.0:
     /usr/local/lib/perl5/5.8.0/i686-linux-thread-multi
     /usr/local/lib/perl5/5.8.0
     /usr/local/lib/perl5/site_perl/5.8.0/i686-linux-thread-multi
     /usr/local/lib/perl5/site_perl/5.8.0
     /usr/local/lib/perl5/site_perl/5.7.3
     /usr/local/lib/perl5/site_perl
     .


Environment for perl v5.8.0:
     HOME=/home/liz
     LANG=en_US
     LANGUAGE (unset)
     LD_LIBRARY_PATH (unset)
     LOGDIR (unset)
     PATH=/usr/local/bin:/bin:/usr/bin:/usr/X11R6/bin:/home/liz/bin
     PERL_BADLANG (unset)
     SHELL=/bin/bash



```
</details>

**Comment by p5pRT** [comment]
### From @tux

On Thu 08 Aug 2002 18&#8203;:06\, Elizabeth Mattijsen \(via RT\) \<perlbug@&#8203;perl\.org> wrote&#8203;:
> \# New Ticket Created by  Elizabeth Mattijsen 
> \# Please include the string&#8203;:  \[perl \#16080\]
> \# in the subject line of all future correspondence about this issue\. 
> \# \<URL&#8203;: http&#8203;://rt\.perl\.org/rt2/Ticket/Display\.html?id=16080 >
> 
> 
> This is a bug report for perl from liz@&#8203;dijkmat\.nl\,
> generated with the help of perlbug 1\.34 running under perl v5\.8\.0\.
> 
> 
> \-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-
> \[Please enter your report here\]
> 
> $ perl \-e '\{my $foo : shared\}'
> Invalid separator character '\}' in attribute list at \-e line 1\, near "$foo 
> : shared"
> syntax error at \-e line 1\, near "$foo : shared"
> Execution of \-e aborted due to compilation errors\.
> 
> 
> Found this problem when trying to do a&#8203;:
> 
>    BEGIN \{our $foo : shared\}
> 
> of course&#8203;:
> 
>    BEGIN \{our $foo : shared;\}
> 
> works ok\.\.\.

Does sound like perl\-1 :\)

\-\- 
H\.Merijn Brand        Amsterdam Perl Mongers \(http&#8203;://amsterdam\.pm\.org/\)
using perl\-5\.6\.1\, 5\.8\.0 & 633 on HP\-UX 10\.20 & 11\.00\, AIX 4\.2\, AIX 4\.3\,
&nbsp;  WinNT 4\, Win2K pro & WinCE 2\.11\.  Smoking perl CORE&#8203;: smokers@&#8203;perl\.org
http&#8203;://archives\.develooper\.com/daily\-build@&#8203;perl\.org/   perl\-qa@&#8203;perl\.org
send smoke reports to&#8203;: smokers\-reports@&#8203;perl\.org\, QA&#8203;: http&#8203;://qa\.perl\.org

**Comment by p5pRT** [comment]
### From @rgs

Elizabeth Mattijsen \(via RT\) wrote&#8203;:
> $ perl \-e '\{my $foo : shared\}'
> Invalid separator character '\}' in attribute list at \-e line 1\, near "$foo 
> : shared"
> syntax error at \-e line 1\, near "$foo : shared"
> Execution of \-e aborted due to compilation errors\.
> 
> 
> Found this problem when trying to do a&#8203;:
> 
>    BEGIN \{our $foo : shared\}
> 
> of course&#8203;:
> 
>    BEGIN \{our $foo : shared;\}
> 
> works ok\.\.\.

Fixed :
Change 17704 by rgs@&#8203;rgs\-home on 2002/08/08 19&#8203;:47&#8203;:44

&nbsp;    Fix bug \#16080 : an attribute list should end at '\}'

Affected files \.\.\.

\.\.\.\.\.\. //depot/perl/t/op/attrs\.t\#10 edit
\.\.\.\.\.\. //depot/perl/toke\.c\#439 edit

Differences \.\.\.

==== //depot/perl/t/op/attrs\.t\#10 \(text\) ====

@&#8203;@&#8203; \-8\,7 \+8\,7 @&#8203;@&#8203;
&nbsp;     require '\./test\.pl';
 \}
 
\-plan tests => 37;
\+plan tests => 39;
 
 $SIG\{\_\_WARN\_\_\} = sub \{ die @&#8203;\_ \};
 
@&#8203;@&#8203; \-50\,6 \+50\,12 @&#8203;@&#8203;
 eval 'my \($x\,$y\) : plugh;';
 like $@&#8203;\, qr/^Invalid SCALAR attribute&#8203;: \["'\]?plugh\["'\]? at/;
 
\+\# bug \#16080
\+eval '\{my $x : plugh\}';
\+like $@&#8203;\, qr/^Invalid SCALAR attribute&#8203;: \["'\]?plugh\["'\]? at/;
\+eval '\{my \($x\,$y\) : plugh\(\}\)\}';
\+like $@&#8203;\, qr/^Invalid SCALAR attribute&#8203;: \["'\]?plugh\\\(\}\\\)\["'\]? at/;
\+
 sub A&#8203;::MODIFY\_SCALAR\_ATTRIBUTES \{ return \}
 eval 'my A $x : plugh;';
 like $@&#8203;\, qr/^SCALAR package attribute may clash with future reserved word&#8203;: \["'\]?plugh\["'\]? at/;

==== //depot/perl/toke\.c\#439 \(text\) ====

@&#8203;@&#8203; \-3049\,7 \+3049\,7 @&#8203;@&#8203;
&nbsp;             break;    /\* require real whitespace or :'s \*/
&nbsp;         \}
&nbsp;         tmp = \(PL\_expect == XOPERATOR ? '=' : '\{'\); /\*'\}\(' for vi \*/
\-        if \(\*s \!= ';' && \*s \!= tmp && \(tmp \!= '=' || \*s \!= '\)'\)\) \{
\+        if \(\*s \!= ';' && \*s \!= '\}' && \*s \!= tmp && \(tmp \!= '=' || \*s \!= '\)'\)\) \{
&nbsp;         char q = \(\(\*s == '\\''\) ? '"' : '\\''\);
&nbsp;         /\* If here for an expression\, and parsed no attrs\, back off\. \*/
&nbsp;         if \(tmp == '=' && \!attrs\) \{

**Comment by p5pRT** [comment]
@gbarr - Status changed from 'new' to 'resolved'
